ryhope/src/storage/updatetree.rs

```diff
@@ -295,52 +399,97 @@ impl<T: Clone + Hash + PartialEq + Eq> From<&UpdateTreeNode<T>> for WorkplanItem
 #[derive(Clone, Serialize, Deserialize)]
 pub struct UpdatePlan<T: Debug + Clone + Hash + Eq> {
     pub t: UpdateTree<T>,
-    ready: Vec<WorkplanItem<T>>,
+    batch_size: usize,
+    anchors: Vec<T>,
 }
 impl<T: Debug + Clone + Hash + Eq> UpdatePlan<T> {
-    fn new(t: UpdateTree<T>) -> Self {
-        let mut r = UpdatePlan {
+    fn new(t: UpdateTree<T>, max_subtree_size: usize) -> Self {
+        UpdatePlan {
+            // Mark all the leaves as being ready to be processed
+            anchors: t
+                .nodes
+                .iter()
+                .filter(|n| n.is_leaf())
+                .map(|n| n.k.clone())
+                .collect(),
             t,
-            ready: Default::default(),
-        };
-
-        // Mark all the leaves as being ready to be processed
-        for n in r.t.nodes.iter() {
-            if n.is_leaf() {
-                r.ready.push(n.into());
-            }
+            batch_size: max_subtree_size,
         }
-
-        r
     }
 
+    /// Mark the given item as having been completed. Its dependent will not be
+    /// generated by the iterator until the item has been marked as completed.
     pub fn done(&mut self, item: &WorkplanItem<T>) -> Result<()> {
         let i = *self
             .t
             .idx
-            .get(&item.k)
+            .get(&item.k())
             .ok_or_else(|| anyhow!("unknwown key"))?;
 
         // May happen when restarting a plan
-        self.ready.retain(|x| x.k != item.k);
+        self.anchors.retain(|k| k != item.k());
 
         // Root node is hardcoded to 0
         if i == 0 {
             self.t.nodes.clear();
         } else {
-            let parent = self.t.nodes[i].parent.unwrap();
-            debug_assert!(self.t.nodes[parent].children.contains(&i));
-            self.t.nodes[parent].children.remove(&i);
-            if self.t.nodes[parent].children.is_empty() {
-                self.ready.push((&self.t.nodes[parent]).into());
+            let parent = self.t.node(i).parent.unwrap();
+            debug_assert!(self.t.node(parent).children.contains(&i));
+            self.t.node_mut(parent).children.remove(&i);
+            if self.t.node(parent).children.is_empty() {
+                self.anchors.push(self.t.nodes[parent].k.clone());
             }
         }
         Ok(())
     }
 
+    /// Indicated whether this workplan has been entirely consumed.
     pub fn completed(&self) -> bool {
         self.t.nodes.is_empty()
     }
+
+    /// Generate a new [`WorkplanItem`] satisfying, if possible, the given batch
+    /// size criterion.
+    ///
+    /// The subtree generated is the highest parent of the next anchor such that
+    /// the subtree it spans satisfies the batch size criterion. Note that if
+    /// the batch size is equal to 1, the behavior is identical to a leaf-first
+    /// traversal of the tree.
```

Comment:
Can we keep the previous API as well by just exporting a method that automatically set this size to 1 ?

Reply:
Yep, `into_workplan` is still there.